uniffi_bindgen/src/bindings/gecko/gen_gecko.rs

```diff
@@ -0,0 +1,355 @@
+            Type::Map(type_) => format!("record<DOMString, {}>", type_webidl(type_)?),
+        })
+    }
+
+    /// Declares a C type in the `extern` declarations.
+    pub fn type_ffi(type_: &FFIType) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            FFIType::Int8 => "int8_t".into(),
+            FFIType::UInt8 => "uint8_t".into(),
+            FFIType::Int16 => "int16_t".into(),
+            FFIType::UInt16 => "uint16_t".into(),
+            FFIType::Int32 => "int32_t".into(),
+            FFIType::UInt32 => "uint32_t".into(),
+            FFIType::Int64 => "int64_t".into(),
+            FFIType::UInt64 => "uint64_t".into(),
+            FFIType::Float32 => "float".into(),
+            FFIType::Float64 => "double".into(),
+            FFIType::RustBuffer => "RustBuffer".into(),
+            FFIType::RustString => "char*".into(),
+            FFIType::RustError => "NativeRustError".into(),
+            FFIType::ForeignStringRef => "const char*".into(),
+        })
+    }
+
+    pub fn ret_type_ffi(type_: &Type) -> Result<String, askama::Error> {
+        let ffi_type = FFIType::from(type_);
+        Ok(type_ffi(&ffi_type)?)
+    }
+
+    /// Declares the type of an argument for the C++ binding.
+    pub fn arg_type_cpp(type_: &Type) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            Type::Int8
+            | Type::UInt8
+            | Type::Int16
+            | Type::UInt16
+            | Type::Int32
+            | Type::UInt32
+            | Type::Int64
+            | Type::UInt64
+            | Type::Float32
+            | Type::Float64
+            | Type::Boolean => type_cpp(type_)?,
+            Type::String => "const nsAString&".into(),
+            Type::Enum(name) => name.into(),
+            Type::Record(name) | Type::Object(name) => format!("const {}&", class_name_cpp(name)?),
+            Type::Error(name) => panic!("[TODO: type_cpp({:?})]", type_),
+            // Nullable objects might be passed as pointers, not sure?
+            Type::Optional(type_) => format!("const Nullable<{}>&", type_cpp(type_)?),
+            Type::Sequence(type_) => format!("const Sequence<{}>&", type_cpp(type_)?),
+            Type::Map(type_) => format!("const Record<nsString, {}>&", type_cpp(type_)?),
+        })
+    }
+
+    pub fn type_cpp(type_: &Type) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            Type::Int8 => "int8_t".into(),
+            Type::UInt8 => "uint8_t".into(),
+            Type::Int16 => "int16_t".into(),
+            Type::UInt16 => "uint16_t".into(),
+            Type::Int32 => "int32_t".into(),
+            Type::UInt32 => "uint32_t".into(),
+            Type::Int64 => "int64_t".into(),
+            Type::UInt64 => "uint64_t".into(),
+            Type::Float32 => "float".into(),
+            Type::Float64 => "double".into(),
+            Type::Boolean => "bool".into(),
+            Type::String => "nsAString".into(),
```

Comment:
This isn’t conceptually right (it should be `nsString`), but it _is_ the type we need to specialize `ViaFfi::Lift` and `Lower`. That might need to go into its own method...

And also, `type_cpp` needs a better name—something that captures this is the type expressed in a type parameter or when declaring a value.